### PR TITLE
Proposal: Support Custom CA Certificate for Karmada Instance in Operator

### PR DESCRIPTION
**Pull request by jabellard**
**What type of PR is this?**

/kind design

**What this PR does / why we need it**:
- Details are contained in the proposal.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: docs/proposals/karmada-operator/custom_ca_cert/README.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+// CustomCertConfig defines the structure for user-provided CA certificates in Custom mode.
+type CustomCertConfig struct {
+   // CACert references a Kubernetes secret containing the CA certificate for the Karmada instance.
+   // The secret must contain the following data keys:
+   // tls.crt: The TLS certificate.
+   // tls.key: The TLS private key.
+   CACert *LocalSecretReference `json:"caCert,omitempty"`
+}
```
Comment (RainbowMango) [inline docs/proposals/karmada-operator/custom_ca_cert/README.md:73]: Is there a possibility to add other fields to customize other certificates?
I don't know, just asking and thinking if there is any interaction with the certification rotation feature in the future.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+// CertificateMode defines the mode for managing certificates for the Karmada control plane.
+type CertificateMode string
+
+const (
+   // Auto mode means the certificate will be automatically generated by the operator.
+   Auto CertificateMode = "Auto"
+   
+   // Custom mode means the user provides their own CA certificate.
+   Custom CertificateMode = "Custom"
+)
```
Comment (RainbowMango) [inline docs/proposals/karmada-operator/custom_ca_cert/README.md:64]: I know certificate rotation is out of the scope of this proposal, just wondering if this mode is extendable enough.
Any thoughts?

**Review by RainbowMango — COMMENTED** [review]
Looks great, thank you @jabellard.

Just want to give another look regarding to certificate rotation, we probably need to reserve extension for that. 
Let me know your thoughts about rotation feature.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+and disaster recovery is essential. Deploying control planes across multiple management clusters that span multiple data centers provides redundancy and resilience in the event of a data center outage. 
+However, when multiple control plane instances must access a shared etcd instance and present a unified API endpoint, using a common CA certificate across instances is necessary. 
+By enabling users to specify a custom CA certificate, this feature ensures that control planes spanning data centers function as a cohesive unit while meeting security and availability standards.
+
+### Architecture Overview
+![architecture-overview](./architecture-overview.png)
+
+### Goals
+
+- Allow users to specify a custom CA certificate for the Karmada control plane.
```
[the remark this answers is not in the document]
Comment (zhzhuang-zju) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:34]: When deploying a Karmada instance, the karmada-operator currently generates three CAs: `KarmadaCertRootCA`, `KarmadaCertFrontProxyCA`, and `KarmadaCertEtcdCA`. Is the custom CA used for the user-defined `KarmadaCertRootCA`?

File: docs/proposals/karmada-operator/custom_ca_cert/README.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+this feature enables multi-data center redundancy for critical Karmada deployments and aligns with organizational policies around disaster recovery (DR). This capability
+is especially important for cases where a managed Karmada control plane is stretched across multiple management clusters that span multiple data centers and shares an underlying etcd instance, requiring the control plane instances to have a unified CA.
+
+## Motivation
+
+In high-availability scenarios where Karmada control planes power mission-critical workloads, ensuring that each managed control plane adheres to strict organizational policies around availability
+and disaster recovery is essential. Deploying control planes across multiple management clusters that span multiple data centers provides redundancy and resilience in the event of a data center outage. 
+However, when multiple control plane instances must access a shared etcd instance and present a unified API endpoint, using a common CA certificate across instances is necessary.
```
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:26]: Yes. That's correct.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:34]: Thanks to @zhzhuang-zju for elaborating a bit more on this. As pointed out, when setting up PKI for the Karmada instance, the operator will generate three CAs: Root/Kubernetes CA, Front Proxy CA, and ETCD CA.  This feature is for providing a custom root CA, which is used to sign the API Server cert, and is also used to verify client certificates.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+- Allow users to specify a custom CA certificate for the Karmada control plane.
+- Ensure that control plane instances deployed across multiple management clusters that span multiple data centers use the same CA, enabling secure and seamless cross-data center communication.
+- Enable operators to align Karmada control plane PKI with organizational policies for high availability and security.
+
+### Non-Goals
+
+- Change the default behavior of the Karmada operator when no custom CA is provided.
+- Alter the process of how control planes are connected to etcd.
+- Address custom CA requirements for other Karmada components.
```
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:42]: This is partly addressed in the comments above, but I'll push a change to be a bit more specific in the language used here.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 1 above)
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:73]: Yes. Absolutely. Right now, this is very minimal, but can be extended in the future.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:26]: Will refine the verbiage to be a bit more specific.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 2 above)
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:64]: I think so. Right now, it's very minimal, but can be extended later on for more use cases. I do need the certificate rotation feature, and eventually want to make a contribution for that as well.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md]: Refined the verbiage to be a bit more specific.

**Comment by jabellard** [comment]
@RainbowMango , @zhzhuang-zju : Thanks for your comments and questions. Just pushed a change to refine this proposal based on your feedback. Just let me know if you have other comments.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+type CertificateConfig struct {
+   // Mode defines how certificates for the Karmada control plane should be managed.
+   // Valid values are "Auto" and "Custom". Defaults to "Auto".
+   // +kubebuilder:validation:Enum=Auto;Custom
+   // +kubebuilder:default=Auto
+   Mode CertificateMode `json:"mode,omitempty"`
+   
+   // Custom contains the configuration for user-provided certificates when in Custom mode.
+   // +optional
+   Custom *CustomCertConfig `json:"custom,omitempty"`
```
Comment (zhzhuang-zju) [inline docs/proposals/karmada-operator/custom_ca_cert/README.md:85]: Just a question, if the CA that can be customized is not limited to `KarmadaCertRootCA` and the API also allows customization of `FrontProxyCA`, how should the `Mode` be used in conjunction with `Custom`? For example, in a scenario where `KarmadaCertRootCA` needs to be customized, but `FrontProxyCA` does not need to be customized?

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md]: OK. I see, the current approach focuses on customizing the certificate for Karmada API Server.
@zhzhuang-zju Do you have any comments?

[the remark this answers is not in the document]
Comment (RainbowMango) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md]: Oh, I see the comments below https://github.com/karmada-io/karmada/pull/5794#discussion_r1835886906.

File: docs/proposals/karmada-operator/custom_ca_cert/README.md (hunk 6 above)
Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:85]: @zhzhuang-zju , good question. Given that the front proxy CA is used exclusively for in-cluster communication, I don't have a use case for customizing it, so right now, the plan is to add support only for customizing the Root/Kubernetes CA. However, if a use case for customizing it does arise, then we can easily add support for. It's a completely independent, self-singed CA: 
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 4436933017513869186 (0x3d932712511dbf82)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=front-proxy-ca
        Validity
            Not Before: Nov  2 17:08:48 2024 GMT
            Not After : Oct 31 17:08:48 2034 GMT
        Subject: CN=front-proxy-ca
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (3072 bit)
                
                ...
```

Comment (zhzhuang-zju) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:85]: Thanks for the explanation. It makes sense. As long as the extensibility is ensured, we can consider supporting more certificate configs when there are relevant use cases in the future.

Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:85]: @RainbowMango , @zhzhuang-zju , any more questions/comments?

Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:85]: Will make name of CA cert more specific. Maybe `KarmadaRootCACert`???

Comment (jabellard) [reply docs/proposals/karmada-operator/custom_ca_cert/README.md:85]: Add `optional` tag to secret ref for Karmada root CA cert as it is optional.

**Comment by jabellard** [comment]
@RainbowMango , I pushed changes. Any other comments?

Also, out of curiosity,  is any of the Karmada maintainers part of the multi cluster SIG? I'm currently at KubeCon and have been talking to some of the members of that group. Looks like they're some overlap in the work that they do.

**Comment by RainbowMango** [comment]
It's still in my queue, I will get back on this ASAP.

> Looks like they're some overlap in the work that they do.

I contributed a lot to that SIG, what kind of work overlap you are talking about?

**Comment by jabellard** [comment]
> It's still in my queue, I will get back on this ASAP.
> 
> > Looks like they're some overlap in the work that they do.
> 
> I contributed a lot to that SIG, what kind of work overlap you are talking about?

Yeah. I was wondering if you've contributed to that SIG and just found out you've contributed to the `Work` API. That's awesome!

**Comment by RainbowMango** [comment]
Hi @jabellard I worked out another approach that doesn't need the `CertificateMode`, please take a look:
```golang

// KarmadaSpec is the specification of the desired behavior of the Karmada.
type KarmadaSpec struct {
	// CustomCertificate specifies the configuration to customize the certificates
	// for Karmada components or control the certificate generation process, such as
	// the algorithm, validity period, etc.
	// Currently, it only supports customizing the CA certificate for limited components.
	// +optional
	CustomCertificate *CustomCertificate `json:"customCertificate,omitempty"`
}

// CustomCertificate holds the configuration for generating the certificate.
type CustomCertificate struct {
	// APIServerCACert references a Kubernetes secret containing the CA certificate
	// for component karmada-apiserver.
	// The secret must contain the following data keys:
	// - tls.crt: The TLS certificate.
	// - tls.key: The TLS private key.
	// If specified, this CA will be used to issue client certificates for
	// all components that access the APIServer as clients.
	// +optional
	APIServerCACert *LocalSecretReference `json:"apiServerCACert,omitempty"`
}
```

I'm concerned about the `CertificateMode` as it might bring ambitions, since `Custom` seems like all certificates will be customized.

**Comment by jabellard** [comment]
@RainbowMango , I think that's a valid concern. Thanks for thinking about  how to remedy it. This design is also a bit simpler.  Will push an update to incorporate it.

**Review by RainbowMango — COMMENTED** [review]
/lgtm
Thank you @jabellard for the effort.

I think it's ready to go after squashing the commits.

**Review by RainbowMango — COMMENTED** [review]
/lgtm
/approve
Thanks.

**Comment by jabellard** [comment]
Hey @RainbowMango , I just submitted the [implementation PR](https://github.com/karmada-io/karmada/pull/5842) for this. Can you please take a look?

**Comment by RainbowMango** [comment]
Sure.  I hope this can be included in the coming release (v1.12.0) which is planed at the end of this month.

**Pull request merged**